File: legacy/python/src/parsers/description_parser.py
"""Parse video descriptions and titles for entities"""
import re
from typing import Dict, List
from urllib.parse import urlparse


class DescriptionParser:
    """Extract entities from video descriptions and titles without LLM"""
    
    # Regex patterns
    GITHUB_REPO_PATTERN = r'github\.com/([a-zA-Z0-9_-]+)/([a-zA-Z0-9_.-]+)'
    URL_PATTERN = r'https?://[^\s<>"{}|\\^`\[\]]+'
    
    def __init__(self):
        self.github_regex = re.compile(self.GITHUB_REPO_PATTERN, re.IGNORECASE)
        self.url_regex = re.compile(self.URL_PATTERN)
    
    def parse(self, title: str, description: str) -> Dict[str, List]:
        """
        Parse title and description for entities
        
        Args:
            title: Video title
            description: Video description
            
        Returns:
            Dictionary with extracted entities
        """
        # Combine title and description for parsing
        combined_text = f"{title}\n{description}"
        
        # Extract GitHub repos
        github_repos = self._extract_github_repos(combined_text)
        
        # Extract all URLs
        all_urls = self._extract_urls(combined_text)
        
        # Filter out GitHub URLs from general websites
        github_domains = {'github.com', 'raw.githubusercontent.com', 'gist.github.com'}
        websites = [
            url for url in all_urls 
            if urlparse(url).netloc.lower() not in github_domains
        ]
        
        return {
            'github_repos': github_repos,
            'websites': websites,
            'topics': [],  # Could add keyword extraction here
            'people': [],  # Could add @mention extraction here
            'key_concepts': [],
            'summary': None
        }
    
    def _extract_github_repos(self, text: str) -> List[Dict]:
        """Extract GitHub repository URLs"""
        repos = []
        seen = set()
        
        matches = self.github_regex.finditer(text)
        for match in matches:
            owner = match.group(1)
            repo_name = match.group(2)
            
            # Clean repo name (remove trailing dots, etc)
            repo_name = repo_name.rstrip('.')
            
            repo_key = f"{owner}/{repo_name}".lower()
            if repo_key not in seen:
                seen.add(repo_key)
                repos.append({
                    'url': f"https://github.com/{owner}/{repo_name}",
                    'owner': owner,
                    'name': repo_name,
                    'full_name': f"{owner}/{repo_name}"
                })
        
        return repos
    
    def _extract_urls(self, text: str) -> List[str]:
        """Extract all URLs from text"""
        urls = []
        seen = set()
        
        matches = self.url_regex.finditer(text)
        for match in matches:
            url = match.group(0)
            
            # Clean up trailing punctuation
            url = url.rstrip('.,;:)]}')
            
            if url.lower() not in seen:
                seen.add(url.lower())
                urls.append(url)
        
        return urls
    
    def extract_entities_for_database(self, parsed_data: Dict) -> List[Dict]:
        """
        Convert parsed data to database entity format
        
        Args:
            parsed_data: Output from parse()
            
        Returns:
            List of entity dictionaries for database storage
        """
        entities = []
        
        # Add GitHub repos
        for repo in parsed_data.get('github_repos', []):
            entities.append({
                'type': 'github_repo',
                'value': repo['full_name'],
                'url': repo['url'],
                'confidence': 100
            })
        
        # Add websites
        for website in parsed_data.get('websites', []):
            entities.append({
                'type': 'website',
                'value': website,
                'url': website,
                'confidence': 100
            })
        
        return entities
    
    def get_tags(self, parsed_data: Dict) -> List[str]:
        """Extract tags from parsed data"""
        tags = []
        
        # Add 'github' tag if repos found
        if parsed_data.get('github_repos'):
            tags.append('github')
        
        # Add 'has-links' tag if websites found
        if parsed_data.get('websites'):
            tags.append('has-links')
        
        return tags

File: src-ts-v2/parsers/DescriptionParser.ts
/**
 * Parse video descriptions and titles for entities without LLM.
 *
 * Deterministic regex-only parser: extracts GitHub repositories and URLs from
 * the combined title+description text. Lifted KEEP-AS-IS from
 * `src-ts/parsers/DescriptionParser.ts` per `docs/PORT_PLAN.md` Wave 3.
 *
 * Canonical Python reference: `legacy/python/src/parsers/description_parser.py`.
 *
 * Known gap (P7 disposition, documented in `docs/MIGRATION_NOTES.md`):
 * `topics` and `people` are returned as empty arrays. The Python parent has
 * the same gap by design — these slots exist so the downstream extractor
 * pipeline can merge in Gemini's output without changing the shape. The
 * regex-only parser never populates them; Gemini does.
 */

import logger from '../utils/logger.js';
import { ValidationError } from '../errors/index.js';

/**
 * Extracted GitHub repository information
 */
export interface GitHubRepo {
  url: string;
  owner: string;
  name: string;
  full_name: string;
}

/**
 * Parsed description result
 */
export interface ParsedDescription {
  github_repos: GitHubRepo[];
  websites: string[];
  topics: string[];
  people: string[];
  key_concepts: string[];
  summary: string | null;
}

/**
 * Database entity format
 */
export interface DatabaseEntity {
  type: string;
  value: string;
  url: string | null;
  confidence: number;
}

/**
 * DescriptionParser class for extracting entities from video descriptions
 */
export class DescriptionParser {
  private readonly githubRegex: RegExp;
  private readonly urlRegex: RegExp;

  /**
   * Regex patterns for extraction
   */
  private static readonly GITHUB_REPO_PATTERN =
    /github\.com\/([a-zA-Z0-9_-]+)\/([a-zA-Z0-9_.-]+)/gi;
  private static readonly URL_PATTERN = /https?:\/\/[^\s<>"{}|\\^`\[\]]+/gi;

  constructor() {
    this.githubRegex = new RegExp(DescriptionParser.GITHUB_REPO_PATTERN.source, 'gi');
    this.urlRegex = new RegExp(DescriptionParser.URL_PATTERN.source, 'gi');
  }

  /**
   * Parse title and description for entities
   *
   * @param title - Video title
   * @param description - Video description
   * @returns Parsed entities
   * @throws {ValidationError} If inputs are invalid
   */
  parse(title: string, description: string): ParsedDescription {
    // Validate inputs
    if (typeof title !== 'string') {
      throw new ValidationError('title must be a string');
    }

    if (typeof description !== 'string') {
      throw new ValidationError('description must be a string');
    }

    try {
      // Combine title and description for parsing
      const combinedText = `${title}\n${description}`;

      // Extract GitHub repos
      const github_repos = this.extractGitHubRepos(combinedText);

      // Extract all URLs
      const allUrls = this.extractUrls(combinedText);

      // Filter out GitHub URLs from general websites
      const githubDomains = new Set(['github.com', 'raw.githubusercontent.com', 'gist.github.com']);

      const websites = allUrls.filter((url) => {
        try {
          const urlObj = new URL(url);
          return !githubDomains.has(urlObj.hostname.toLowerCase());
        } catch {
          return false;
        }
      });

      logger.debug(
        {
          github_repos_count: github_repos.length,
          websites_count: websites.length,
        },
        'Parsed description'
      );

      return {
        github_repos,
        websites,
        topics: [],
        people: [],
        key_concepts: [],
        summary: null,
      };
    } catch (error) {
      if (error instanceof ValidationError) {
        throw error;
      }

      logger.error(
        {
          error: error instanceof Error ? error.message : String(error),
        },
        'Failed to parse description'
      );

      throw new ValidationError('Failed to parse description', {
        cause: error,
      });
    }
  }

  /**
   * Extract GitHub repository URLs
   *
   * @param text - Text to extract from
   * @returns Array of GitHub repo objects
   */
  private extractGitHubRepos(text: string): GitHubRepo[] {
    const repos: GitHubRepo[] = [];
    const seen = new Set<string>();

    // Reset regex state
    this.githubRegex.lastIndex = 0;

    let match: RegExpExecArray | null;
    while ((match = this.githubRegex.exec(text)) !== null) {
      const owner = match[1];
      let repoName = match[2];

      // Clean repo name (remove trailing dots, etc)
      repoName = repoName.replace(/\.+$/, '');

      const repoKey = `${owner}/${repoName}`.toLowerCase();

      if (!seen.has(repoKey)) {
        seen.add(repoKey);
        repos.push({
          url: `https://github.com/${owner}/${repoName}`,
          owner,
          name: repoName,
          full_name: `${owner}/${repoName}`,
        });
      }
    }

    return repos;
  }

  /**
   * Extract all URLs from text
   *
   * @param text - Text to extract from
   * @returns Array of URLs
   */
  private extractUrls(text: string): string[] {
    const urls: string[] = [];
    const seen = new Set<string>();

    // Reset regex state
    this.urlRegex.lastIndex = 0;

    let match: RegExpExecArray | null;
    while ((match = this.urlRegex.exec(text)) !== null) {
      let url = match[0];

      // Clean up trailing punctuation
      url = url.replace(/[.,;:)\]}]+$/, '');

      const urlLower = url.toLowerCase();
      if (!seen.has(urlLower)) {
        seen.add(urlLower);
        urls.push(url);
      }
    }

    return urls;
  }

  /**
   * Convert parsed data to database entity format
   *
   * @param parsedData - Output from parse()
   * @returns List of entity dictionaries for database storage
   * @throws {ValidationError} If parsed data is invalid
   */
  extractEntitiesForDatabase(parsedData: ParsedDescription): DatabaseEntity[] {
    if (!parsedData || typeof parsedData !== 'object') {
      throw new ValidationError('parsedData must be an object');
    }

    const entities: DatabaseEntity[] = [];

    try {
      // Add GitHub repos
      if (Array.isArray(parsedData.github_repos)) {
        for (const repo of parsedData.github_repos) {
          entities.push({
            type: 'github_repo',
            value: repo.full_name,
            url: repo.url,
            confidence: 100,
          });
        }
      }

      // Add websites
      if (Array.isArray(parsedData.websites)) {
        for (const website of parsedData.websites) {
          entities.push({
            type: 'website',
            value: website,
            url: website,
            confidence: 100,
          });
        }
      }

      return entities;
    } catch (error) {
      logger.error(
        {
          error: error instanceof Error ? error.message : String(error),
        },
        'Failed to extract entities for database'
      );

      throw new ValidationError('Failed to extract entities for database', {
        cause: error,
      });
    }
  }

  /**
   * Extract tags from parsed data
   *
   * @param parsedData - Output from parse()
   * @returns List of tags
   * @throws {ValidationError} If parsed data is invalid
   */
  getTags(parsedData: ParsedDescription): string[] {
    if (!parsedData || typeof parsedData !== 'object') {
      throw new ValidationError('parsedData must be an object');
    }

    const tags: string[] = [];

    try {
      // Add 'github' tag if repos found
      if (Array.isArray(parsedData.github_repos) && parsedData.github_repos.length > 0) {
        tags.push('github');
      }

      // Add 'has-links' tag if websites found
      if (Array.isArray(parsedData.websites) && parsedData.websites.length > 0) {
        tags.push('has-links');
      }

      return tags;
    } catch (error) {
      logger.error(
        {
          error: error instanceof Error ? error.message : String(error),
        },
        'Failed to extract tags'
      );

      return [];
    }
  }
}
